### DISCUSSION TITLE
Any plan to restructure LanguageServer part to use BaseProtocol?

### DISCUSSION THREAD

**Discussion by 616b2f** [dotnet/roslyn#71340] [General] [answered]
Hi,

I wanted to create a small POC to implement Build Server Protocol for dotnet (https://build-server-protocol.github.io/docs/specification). To avoid to copy paste a lot of the LSP code I would like to use the BaseProtocol (which is currently experimental as it seems, https://microsoft.github.io/language-server-protocol/specifications/base/0.9/specification/).

Are there any plans or maybe already an ongoing work to extract the Base Protocol portion of the LSP implementation as an nuget or similar?

Thank you in Advance!

**Comment by CyrusNajmabadi** [comment, marked as answer]
There are no plans on doing any work here in roslyn in this space.  Could you share some information on what hte value would be?  Thanks!

**Reply by 616b2f** [reply]
On the official LSP site there is an afford to create an Base Protocol, that can be used by other LSP like Protocols like Build Server Protocol and by LSP itself and maybe others in the future. The benefit is to have this Base Protocol in the dotnet ecosystem so that dotnet tooling in the community can be based on it. The thing is that when we already have an LSP implementation there is not that much effort to refactor it to be based on Base Protocol (at least it's how I currently understand it) because it's derived from LSP. But maybe someone that implements the LSP here can give his opinion on that, I would really appreciate that.

**Reply by CyrusNajmabadi** [reply]
Still not sure what the benefit would be.  :)  

What changes would you be asking roslyn to make, and what would that enable?

**Reply by 616b2f** [reply]
Just saw that LSP version 3.18 uses the Base Protocol already (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification/#baseProtocol). So it will be needed here in the future, I guess (not sure which version of LSP is currently implemented here).

> What changes would you be asking roslyn to make, and what would that enable?

- Extract the parts that are Base Protocol relevant to an separate project that can be build and published as Nuget Package so it can be reused in other projects.
- this would enable others to easier build clients or servers that want to implement protocols that are based on Base Protocol, such as LSP or DAP or BSP using the C# language.

**Reply by 616b2f** [reply]
Also found this https://github.com/dotnet/roslyn/tree/main/src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework

It seems kind of related but this seems specific to LSP.

**Reply by CyrusNajmabadi** [reply]
> Just saw that LSP version 3.18 uses the Base Protocol already (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification/#baseProtocol).

That's simply describing the format of lsp requests.

**Reply by CyrusNajmabadi** [reply]
> Extract the parts that are Base Protocol relevant to an separate project that can be build and published as Nuget Package so it can be reused in other projects.

I don't understand why that would be needed or beneficial. We already speak a superset here.  And you can already reference our source and packages.

**Reply by CyrusNajmabadi** [reply]
> this would enable others to easier build clients or servers that want to implement protocols that are based on Base Protocol, such as LSP or DAP or BSP using the C# language.

This is a non-goal for roslyn. Roslyn is not a system to build generic language servers. We're very specifically and intentionally the system to build language servers for the .net ecosystem (c#, vb, xaml, etc). 

You are welcome to use our source and libs for whatever purpose you want. But we won't be creating generic libs like what is described above. Thanks!

**Reply by 616b2f** [reply]
Fair enough, thank you for taking the time :)

**Reply by CyrusNajmabadi** [reply]
yw :)
